File: problem2.java
//Time Complexity: O(n*m), where n is the length of the queries array and m is the length of the pattern string.
//Space Complexity: O(n), where n is the length of the queries array.
//ran on leetcode: yes
//check if each character in the string matches the pattern character, and if a character in the string is uppercase and doesn't match the pattern, or if the pattern character is uppercase and doesn't match the string character, then returns false. 

class Solution {
    public List<Boolean> camelMatch(String[] queries, String pattern) {
        List<Boolean> res = new ArrayList<>();

        for(String str : queries)
        {
            int i=0 , j = 0;
            while(true)
            {
                if(i==str.length() && j==pattern.length())
                {
                    res.add(true);
                    break;
                }
                if(j==pattern.length())
                {
                    boolean isValid = true;
                    for(int k=i;k<str.length();k++)
                    {
                        if(Character.isUpperCase(str.charAt(k)))
                        {
                            isValid = false;
                            break;
                        }
                    }
                    res.add(isValid);
                    break;
                    
                }
                if(i==str.length())
                {
                    res.add(false);
                    break;
                }
                char a = str.charAt(i);
                char b = pattern.charAt(j);
                if(a==b)
                {
                    i++;
                    j++;
                }
                else if(Character.isUpperCase(a))
                {
                    res.add(false);
                    break;
                }
                else if(Character.isUpperCase(b))
                {
                    i++;
                }
                else
                    i++;
            }
        }
        return res;
    }
}

File: Problem2.py
class Solution:
    def camelMatch(self, queries: List[str], pattern: str) -> List[bool]:
        # Time complexity: O(N*M), where N is the number of queries and M is the maximum length of a query.
        # Space complexity: O(N), where N is the number of queries.
        def isMatch(query, pattern):
            i, j = 0, 0  # Initialize pointers for query and pattern
            while i < len(query) and j < len(pattern):
                if query[i] == pattern[j]:
                    i += 1  # Move to the next character in query
                    j += 1  # Move to the next character in pattern
                elif query[i].isupper():
                    return False  # If an uppercase character in query doesn't match, it's not a valid match
                else:
                    i += 1  # Continue comparing the next character in query

            # After the loop, if j == len(pattern), it means pattern is completely matched.
            # Now, check if there are any uppercase letters in the remaining part of query.
            if j == len(pattern):
                return not any(char.isupper() for char in query[i:])

            # If j != len(pattern), it means the pattern is not fully matched
            return False

        result = []  # Initialize an empty list to store the results

        for query in queries:
            # Check each query against the pattern and append the result to the list
            result.append(isMatch(query, pattern))

        return result  # Return the list of boolean results
